deepchecks/nlp/text_data.py

```diff
@@ -65,210 +62,116 @@ class TextData:
+            properties: t.Optional[t.Union[pd.DataFrame]] = None,
     ):
         # Require explicitly setting task type if label is provided
         if task_type in [None, 'other']:
             if label is not None:
-                if isinstance(label, t.Sequence):
-                    if pd.notnull(label).any():
-                        raise DeepchecksValueError('task_type must be set when label is provided')
-                else:
-                    raise DeepchecksValueError('task_type must be set when label is provided')
-
+                raise DeepchecksValueError('task_type must be set when label is provided')
             self._task_type = TaskType.OTHER
         elif task_type == 'text_classification':
             self._task_type = TaskType.TEXT_CLASSIFICATION
         elif task_type == 'token_classification':
+            if tokenized_text is None:
+                raise DeepchecksValueError('tokenized_text must be provided for token_classification task type')
+            validate_tokenized_text(tokenized_text)
+            modified = [[str(token) for token in tokens_per_sample] for tokens_per_sample in tokenized_text]
+            self._tokenized_text = np.asarray(modified)
             self._task_type = TaskType.TOKEN_CLASSIFICATION
         else:
             raise DeepchecksNotSupportedError(f'task_type {task_type} is not supported, must be one of '
                                               f'text_classification, token_classification, other')
 
         if raw_text is None and tokenized_text is None:
-            raise DeepchecksValueError('raw_text and tokenized_text cannot both be None')
+            raise DeepchecksValueError('Either raw_text or tokenized_text must be provided')
         elif raw_text is None:
-            self._validate_tokenized_text(tokenized_text)
-            self._tokenized_text = list(tokenized_text)
-            self._text = [' '.join(tokens) for tokens in tokenized_text]
-        elif tokenized_text is None:
-            self._validate_text(raw_text)
-            self._text = list(raw_text)
-            if self._task_type == TaskType.TOKEN_CLASSIFICATION:
-                self._tokenized_text = [sample.split() for sample in self._text]
-            else:
-                self._tokenized_text = None
-        else:
-            self._validate_text(raw_text)
-            self._validate_tokenized_text(tokenized_text)
-            self._text, self._tokenized_text = list(raw_text), list(tokenized_text)
-            if len(raw_text) != len(tokenized_text):
-                raise DeepchecksValueError('raw_text and tokenized_text must have the same length')
-
-        if index is None:
-            self.index = list(range(len(raw_text)))
-        elif len(index) != len(raw_text):
-            raise DeepchecksValueError('index must be the same length as raw_text')
+            self._text = np.asarray([' '.join(tokens) for tokens in tokenized_text])  # Revisit this decision
```

Comment:
Why did you dump the whole elif else block there?

Reply:
Rearranged the order of stuff so not sure what you mean specifically but what i did do, is to add validation that if you selected "token classification" you must provide the tokenized text

Reply:
You changed the if-else block of "what if there's only raw_text / tokenized_text". I do like the changes (I agree it's better to say "give me tokenized text" if the task is token classification).
But, if I understand correctly, I see the following caveat:

If the user supplied tokenized_text but not raw_text, and specified (incorrectly) that the task is text classification, your code breaks (tokenized_text is not None, but also not validated).

Maybe there are more, I hated that if-else block but it was necessary at the time (though I think you covered all other cases)

Reply:
If you selected text classification, we only care about the format of the raw_text + that the label is in the relevant format, which is validated.
Specifically, in your case either you will get an exception for bad prediction/label format or we dont care and all checks will work fine with the raw text

Reply:
👍